File: tests/unit/split-dns.test.ts
import { describe, expect, it } from 'vitest'
import { parseProxyProcesses } from '../../src/main/tun/proxy-discovery'
import { selectFakeIPRange, fakeIPv6Range } from '../../src/main/tun/split-dns'
import { tunConfig } from '../../src/main/tun/tun-config'
import { settingsSchema, captureDomainsSchema } from '../../src/shared/contracts/model'

describe('Fake IP TUN coexistence', () => {
    it('normalizes and persists bounded DNS names while keeping old settings compatible', () => {
        expect(
            captureDomainsSchema.parse(['Example.COM.', '*.example.com', 'api.example.net'])
        ).toEqual(['example.com', 'api.example.net'])
        expect(settingsSchema.parse({ tun: { socksPort: 0 } }).tun.captureDomains).toEqual([])
        expect(() => captureDomainsSchema.parse(Array(101).fill('example.com'))).toThrow()
    })
    it.each([
        'https://example.com',
        'example.com:443',
        'example.com/path',
        'a..com',
        '-a.com',
        'a-.com',
        '127.0.0.1',
        '*',
        'com\nremove other',
        'a'.repeat(64) + '.com'
    ])('rejects invalid capture domain %s', (value) => {
        expect(() => captureDomainsSchema.parse([value])).toThrow()
    })
    it('recognizes external cores without matching Fluxy or paths containing a proxy name', () => {
        expect(
            parseProxyProcesses(`
123 /Applications/Clash Verge.app/Contents/MacOS/verge-mihomo
124 /usr/local/bin/sing-box
125 /Applications/Fluxy.app/Contents/Resources/core/fluxy-core
126 /tmp/sing-box/not-a-proxy
${process.pid} /tmp/sing-box
`)
        ).toEqual([
            { pid: 123, name: 'Mihomo' },
            { pid: 124, name: 'sing-box' }
        ])
    })
    it('avoids an existing proxy pool and LAN routes without treating default routes as ownership', () => {
        expect(selectFakeIPRange('0/1 utun1024\n128.0/1 utun1024\n198.18/16 utun1024')).toBe(
            '198.19.0.0/16'
        )
        expect(selectFakeIPRange('198.18/15 utun1024')).toBe('100.127.0.0/16')
        expect(selectFakeIPRange('198.18/15 utun1024\n100.64/10 en0')).toBe('172.30.0.0/16')
        expect(() => selectFakeIPRange('198.18/15 utun1024\n100.64/10 en0\n172.16/12 en1')).toThrow(
            'No unused'
        )
        expect(selectFakeIPRange('198.19.5.1 link#9')).toBe('100.127.0.0/16')
    })
    it('routes only owned Fake IPs and DNS, preserves native DNS until helper activation, and leaves egress unbound', () => {
        const settings = settingsSchema.parse({}).tun
        const config = tunConfig({
            settings,
            splitDNS: {
                server: '192.168.1.1',
                ipv4Range: '198.19.0.0/16',
                domains: ['fakeip.fluxy.test']
            },
            bridgePort: 18001,
            egressPort: 18002,
            password: 'test',
            interfaceName: 'utun2345',
            egressInterface: ''
        })
        expect(config.inbounds[0]).toMatchObject({
            dns_mode: 'disabled',
            route_address: ['198.19.0.0/16', fakeIPv6Range, '172.31.255.2/32']
        })
        expect(config.outbounds[0]).toEqual({ type: 'direct', tag: 'direct' })
        expect(config.route.rules[0]).toEqual({
            inbound: ['egress'],
            action: 'route',
            outbound: 'direct'
        })
        expect(config.route.rules[1]).toEqual({
            inbound: ['capture'],
            port: 53,
            action: 'hijack-dns'
        })
        expect(config.dns.servers[0]).toEqual({ type: 'udp', tag: 'local', server: '192.168.1.1' })
        expect(config.dns.rules?.[0]).toMatchObject({ domain_suffix: ['fakeip.fluxy.test'] })
        expect(settings.socksPort).toBe(0)
    })
})

File: tools/helper/split_dns_test.go
package main

import (
	"runtime"
	"testing"
)

func TestSplitDNSValidation(t *testing.T) {
	p := tunParams{BridgePort: 6060, EgressPort: 6061, Password: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
		InterfaceName: "utun2345", SplitDNS: &splitDNSParams{IPv4Range: "198.19.0.0/16", Server: "192.168.1.1"}}
	if runtime.GOOS != "darwin" {
		if p.SplitDNS.validate() == nil {
			t.Fatal("accepted unsupported platform")
		}
		return
	}
	if err := p.validate(); err != nil {
		t.Fatal(err)
	}
	for _, test := range []struct {
		name   string
		change func(*tunParams)
	}{
		{"default route", func(p *tunParams) { p.SplitDNS.IPv4Range = "0.0.0.0/0" }},
		{"DNS loop", func(p *tunParams) { p.SplitDNS.Server = splitDNSAddress }},
		{"fake IPv4 loop", func(p *tunParams) { p.SplitDNS.Server = "198.19.1.1" }},
		{"fake IPv6 loop", func(p *tunParams) { p.SplitDNS.Server = "fd7a:115c:a1e0::2" }},
		{"server injection", func(p *tunParams) { p.SplitDNS.Server = "127.0.0.1\nremove other" }},
		{"SOCKS override", func(p *tunParams) { p.SocksPort = 7897 }},
		{"route override", func(p *tunParams) { p.RouteCIDRs = []string{"0.0.0.0/0"} }},
		{"interface override", func(p *tunParams) { p.EgressInterface = "en0" }},
	} {
		t.Run(test.name, func(t *testing.T) {
			next := p
			dns := *p.SplitDNS
			next.SplitDNS = &dns
			test.change(&next)
			if next.validate() == nil {
				t.Fatal("accepted unsafe split DNS profile")
			}
		})
	}
}

func TestSplitDNSConfig(t *testing.T) {
	p := validParams()
	p.SocksPort = 0
	p.RouteCIDRs = nil
	p.SplitDNS = &splitDNSParams{IPv4Range: "198.19.0.0/16", Server: "192.168.1.1"}
	c := config(p)
	inbound := c["inbounds"].([]any)[0].(map[string]any)
	routes := inbound["route_address"].([]string)
	if len(routes) != 3 || routes[0] != p.SplitDNS.IPv4Range || routes[1] != fakeIPv6Range || routes[2] != splitDNSAddress+"/32" {
		t.Fatalf("unexpected capture routes: %v", routes)
	}
	outbound := c["outbounds"].([]any)[0].(map[string]any)
	if _, bound := outbound["bind_interface"]; bound || outbound["type"] != "direct" {
		t.Fatal("real egress does not follow existing routes")
	}
	dns := c["dns"].(map[string]any)
	if dns["servers"].([]any)[0].(map[string]any)["server"] != p.SplitDNS.Server {
		t.Fatal("lost original upstream DNS")
	}
	if c["route"].(map[string]any)["rules"].([]any)[1].(map[string]any)["action"] != "hijack-dns" {
		t.Fatal("TUN DNS does not reach core resolver")
	}
}

func TestCaptureDomains(t *testing.T) {
	for _, domains := range [][]string{nil, {"example.com", "api.example.net"}} {
		if err := validateCaptureDomains(domains); err != nil {
			t.Fatal(err)
		}
	}
	for _, domain := range []string{"", ".", "*", "*.example.com", "https://example.com", "example.com:443", "a..com", "example.com\nremove other", "127.0.0.1", "Example.com"} {
		t.Run(domain, func(t *testing.T) {
			if validateCaptureDomains([]string{domain}) == nil {
				t.Fatal("accepted unsafe resolver domain")
			}
		})
	}
	p := validParams()
	p.SplitDNS = &splitDNSParams{IPv4Range: "198.19.0.0/16", Server: "192.168.1.1", Domains: []string{"example.com"}}
	rules := config(p)["dns"].(map[string]any)["rules"].([]any)
	if rules[0].(map[string]any)["domain_suffix"].([]string)[0] != "example.com" {
		t.Fatal("DNS scope lost")
	}
}
